### ISSUE TITLE
Message buffering is not working when network disconnected

### ISSUE THREAD

**Issue by RAJAP1** [rsyslog/rsyslog#5005]
Hi,

I need to collect the syslog messages from the network devices/windows (udp/tcp) and forward the syslogs to centralized server.

Windows, firewall, switches (LAN) (upd)  --> Forwarder (DMZ)  (CentOS Imudp & omrelp) --> Centralized server (WAN) (CentOS imrelp)

It is working perfectly under normal condition.

The requirement is that the forwarder should receive the syslogs from network devices and buffer it when the centralized receiver is not reachable. It should re-forward again the connectivity resumes back.

For this i have configured the forwarder action queues with omrelp. 

To simulate the sinario, First I stopped the Centralized server's rsyslog service and triggered some logs in the firewall. Once i restarted the service after 1 hour, i can able to see the logs coming to the centralized server immediately which i triggered.

But when i disconnect the network in the firewall,  i could not see any log messages which i triggered after the resume the connection.

I was getting the below error message. 

2022-10-22 07:14:53	Syslog.Error	x.x.x.x	2022-10-21T20:13:17.620820-04:00 localhost rsyslogd[318980]: omrelp[x.x.x.x:514]: error 'error opening connection to remote peer', object  'conn to srvr x.x.x.x:514' - action may not work as intended [v8.2102.0-10.el8 try https://www.rsyslog.com/e/2353 ] 

Is this expected result when network is disconnected? or did i miss something in configuration?

Also, please find the below client and server configuration

**Server Config:**
module(load="imrelp")
input(type="imrelp" port="514" maxDataSize="10k" keepAlive="on")

**Client Config:**
module(load="imudp") # needs to be done just once
input(type="imudp" port="514")
module(load="omrelp")
action(
        type="omrelp"
        target="Centralized server IP"
        port="514"
        timeout="5"
        conn.timeout="3"
        RebindInterval="30"
        tls="off"
        action.reportSuspension="on"
        action.reportSuspensionContinuation="on"
        action.resumeRetryCount="-1"
        action.resumeInterval="10"
        queue.spoolDirectory="/var/spool/rsyslog"
        queue.filename="test-log-spool"
        queue.type="Disk"
        queue.saveonshutdown="on"
)

**Comment by davidelang** [comment]
please enable impstats so that we can see what's happening with the queues.

David Lang

**Comment by RAJAP1** [comment]
[impstats.txt](https://github.com/rsyslog/rsyslog/files/9859083/impstats.txt)

Hi , I have attached the impstats log. Could you please check for the queue issue.

I disconnected the network between client and server & resumed the connectivity respectively Oct 25 07:34:33 & 10:54:37 as per the logs timestamp. 

Regards,
Raja P

**Comment by rgerhards** [comment]
Mhhh... the queue looks ok. It grows and sticks to the messages. 

The question is what happens after the network connect/disconnect. Can this cause some IP addresses to change? This could cause issues.

If @davidelang does not have a better idea than I, I would suggest to create a debug log. Instructions: https://www.rsyslog.com/doc/v8-stable/troubleshooting/howtodebug.html

Be sure that the problem in question occurs at least once during the debug run. Once this is done, please post the debug log (e.g. via pastbin or github). After that, we can review the contents and check how exactly rsyslog processing worked.

**Comment by RAJAP1** [comment]
Hi, 

I have tested again with the same scenario, and i found the that lost messages are triggered immediately after the network disconnection. 

Is this due to connection-timeout setting the action?  Because, i could able to received the other messages which i triggered after sometime during the network disconnection.

What would be the recommended settings? 

action(
        type="omrelp"
        target="172.40.0.2"
        port="514"
        timeout="5"
        conn.timeout="3"
        RebindInterval="30"
        tls="off"

        action.reportSuspension="on"
        action.reportSuspensionContinuation="on"
        action.resumeRetryCount="-1"
        action.resumeInterval="10"
        queue.spoolDirectory="/var/spool/rsyslog"
        queue.filename="test-log-spool"
        queue.type="LinkedList"
        #queue.type="Disk"
        queue.saveonshutdown="on"

**Comment by davidelang** [comment]
are you sure they aren't just being delivered out of order?

logs spooled to the disk will get delivered later than newer logs that arrive 
once the connection is back up (the new logs stay in the memory queue and can be 
delivered much faster than the logs that need to be retrieved from disk)

David Lang

**Comment by davidelang** [comment]
Also, there is one known bug with RELP where one message can be lost. I don't 
remember the details, but there hasn't been enough interest to prioritize fixing 
it. But it's only one message, requiring specific timing.

David Lang

**Comment by rgerhards** [comment]
> Also, there is one known bug with RELP where one message can be lost.

It can be more than one message, up to a RELP window. It happens when

1. client connection to server breaks
2. client is shutdown or restarted

The reason is that the current unacked part of the relp window is only held in main memory for recovery. We had several questions from enterprise customers about this, none wanted to fund a fix, so I concluded that a fix is not worth it (as several enterprises finally considered it irrelevant).

**Comment by rgerhards** [comment]
Apologies for the very late reply. This is probably no longer relevant for the original reporter, but I am adding the answer for people who find this issue through search. We do not lightly close old issues just because of age; we are working through the backlog and closing them when the old thread has a clear answer or a known limitation has already been identified.

The older discussion narrowed this down to RELP behavior around a broken connection and client shutdown/restart. The rsyslog action/disk queue can persist messages that have not yet been handed to RELP, but the RELP protocol window contains messages that have been sent and are waiting for acknowledgement. That unacknowledged window is in librelp/connection state, not in the rsyslog action queue, so a restart during a broken connection can lose up to the RELP window. Disk-spooled messages may also be delivered later than newer in-memory traffic, so ordering can look surprising after recovery.

This was previously identified as requiring coordinated librelp and rsyslog changes, and was not planned in the old thread. I am closing this as answered/known limitation. If someone wants to work on persistent RELP unacknowledged-window handling, please open a new focused feature/bug issue or PR against current rsyslog/librelp with the exact recovery semantics and tests proposed.
